source-snowflake/capture.go

```diff
@@ -0,0 +1,650 @@
+			needsDeletion = append(needsDeletion, stagingTable.Name)
+			continue
+		}
+		var sourceTable, sourceTableKnown = idToSourceTable[uniqueID]
+		if !sourceTableKnown {
+			log.WithField("name", stagingTable.Name).Debug("will not prune staging table of unknown binding")
+			continue
+		}
+		var streamState = c.State.Streams[sourceTable]
+		if int(sequenceNumber) < streamState.SequenceNumber {
+			log.WithField("name", stagingTable.Name).Debug("will prune staging table which has been fully captured")
+			needsDeletion = append(needsDeletion, stagingTable.Name)
+			continue
+		}
+
+		// Since we're already enumerating staging tables and parsing their names, we also
+		// want to keep track of the highest staging table index observed for later. Using
+		// greater-or-equal comparison here ensures that zero works consistently.
+		if int(sequenceNumber) >= highestStaged[sourceTable] {
+			highestStaged[sourceTable] = int(sequenceNumber)
+		}
+	}
+
+	// Actually perform deletion of no-longer-needed staging tables.
+	for _, deleteTable := range needsDeletion {
+		log.WithField("table", deleteTable).Info("pruning staging table")
+		var dropTableQuery = fmt.Sprintf("DROP TABLE IF EXISTS %s;", snowflakeObject{c.Config.Advanced.FlowSchema, deleteTable}.QuotedName())
+		if _, err := c.DB.ExecContext(ctx, dropTableQuery); err != nil {
+			return nil, fmt.Errorf("error dropping staging table %q: %w", deleteTable, err)
+		}
+	}
+	return highestStaged, nil
+}
+
+func (c *capture) captureStagingTable(ctx context.Context, table snowflakeObject, orderColumns []string) error {
+	var streamState = c.State.Streams[table]
+	var stagingName = stagingTableName(c.Config, c.Name, table, streamState.SequenceNumber)
+	var logEntry = log.WithFields(log.Fields{
+		"table": table.String(),
+		"seq":   streamState.SequenceNumber,
+	})
+
+	// Ensure that the staging table exists by creating it if necessary. Since
+	// the creation is done IF NOT EXISTS this is a no-op if it already does.
+	if _, err := createStagingTable(ctx, c.Config, c.DB, c.Name, table, streamState.SequenceNumber); err != nil {
+		return err
+	}
+
+	// Generate a query to read the staging table contents.
+	var quotedOrderColumns []string
+	for _, orderColumn := range orderColumns {
+		quotedOrderColumns = append(quotedOrderColumns, quoteSnowflakeIdentifier(orderColumn))
+	}
+	var queryBuf = new(strings.Builder)
+	fmt.Fprintf(queryBuf, "SELECT * FROM %s", stagingName.QuotedName())
+	if len(quotedOrderColumns) > 0 {
+		fmt.Fprintf(queryBuf, " ORDER BY %s", strings.Join(quotedOrderColumns, ", "))
+	}
+	fmt.Fprintf(queryBuf, " LIMIT NULL OFFSET %d", streamState.TableOffset)
+	var readStagingTableQuery = queryBuf.String()
+
+	// Read contents of staging table and output to Flow.
+	logEntry.WithField("query", readStagingTableQuery).Info("querying staging table")
+	rows, err := c.DB.QueryContext(ctx, readStagingTableQuery)
```

Comment:
Definitely test out the streaming JSON downloader option, which we implemented for `materialize-snowflake` load queries, to prevent extremely high memory usage & OOM kills: https://github.com/estuary/connectors/pull/745. FYI I pulled this branch down and threw together a docker file to get a capture working, and am seeing memory usage very near 1 GB for the connector when capturing a modestly sized table.

Also, the driver has some weirdness around capturing numbers, see https://pkg.go.dev/github.com/snowflakedb/gosnowflake#hdr-Supported_Data_Types and the `WithHigherPrecision` option. It might not do this if using the streaming JSON downloader, but as it is I noticed the capture was capturing integers as strings.

Reply:
I've modified the capture to use streaming JSON here, and added proper value translation plumbing to parse stringified numeric and boolean columns back into the appropriate JSON types.